src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonTypeInfo.cs

```diff
@@ -191,24 +191,29 @@ internal JsonTypeInfo(Type type, JsonConverter converter, Type runtimeType, Json
             {
                 case ConverterStrategy.Object:
                     {
+                        const BindingFlags bindingFlags =
+                            BindingFlags.Instance |
+                            BindingFlags.Public |
+                            BindingFlags.NonPublic |
+                            BindingFlags.DeclaredOnly;
+
                         CreateObject = Options.MemberAccessorStrategy.CreateConstructor(type);
-                        Dictionary<string, JsonPropertyInfo> cache = new Dictionary<string, JsonPropertyInfo>(
-                            Options.PropertyNameCaseInsensitive
-                                ? StringComparer.OrdinalIgnoreCase
-                                : StringComparer.Ordinal);
 
                         Dictionary<string, MemberInfo>? ignoredMembers = null;
 
+                        PropertyInfo[] properties = type.GetProperties(bindingFlags);
+
+                        // PropertyCache is not accessed by other threads until the current JsonTypeInfo instance
+                        //  is finished initializing and added to the cache on JsonSerializerOptions.
+                        // Default 'capacity' to the common non-polymorphic + property case.
+                        PropertyCache = new JsonPropertyDictionary<JsonPropertyInfo>(Options.PropertyNameCaseInsensitive, capacity: properties.Length);
+
                         // We start from the most derived type.
-                        for (Type? currentType = type; currentType != null; currentType = currentType.BaseType)
-                        {
-                            const BindingFlags bindingFlags =
-                                BindingFlags.Instance |
-                                BindingFlags.Public |
-                                BindingFlags.NonPublic |
-                                BindingFlags.DeclaredOnly;
+                        Type? currentType = type;
 
-                            foreach (PropertyInfo propertyInfo in currentType.GetProperties(bindingFlags))
+                        while (true)
```

Comment:
Why's an infinite `while` loop + `break` approach preferred to the `for` loop?

Reply:
Before the loop, the initial count of properties via GetProperties() is obtained to initialize the cache with the best guess for the count of properties. Moving to a `while` avoids a redundant call to GetProperties().